### PR TITLE
feat: added markdown support for the comments

### PR DESCRIPTION
**Pull request by rebelchris**
Added markdown powered comments.
The user is able to type markdown and we will parse it into html on the API side.

The user should also be able to edit and see the markdown again.

Out of scope:
Copy function for code blocks

Screenshots:

Edit comment:
Screenshot 2021-11-26 at 15 13 46

Rendered comment:
Screenshot 2021-11-26 at 15 13 39

Other elements:
Screenshot 2021-11-26 at 15 13 54

Other elements 2:
Screenshot 2021-11-26 at 15 14 01

DD-210 #done

### CODE REVIEW

File: packages/shared/src/components/comments/comments.module.css [hunk 1]
```diff
@@ -1,6 +1,147 @@
...
+    .javascript .xml,
+    .tex .hljs-formula,
+    .xml .javascript,
+    .xml .vbscript,
+    .xml .css,
+    .xml .hljs-cdata)) {
+      opacity: 0.5;
+    }
+  }
+}
```
Comment (sshanzel) [inline packages/shared/src/components/comments/comments.module.css:147]: just needs another line for eof

File: packages/shared/src/components/comments/CommentMarkdown.tsx [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import React, { ReactElement } from 'react';
```
Comment (sshanzel) [inline packages/shared/src/components/comments/CommentMarkdown.tsx:1]: In time we might have another place where we will utilize html/markdown contents, so I think it is better to make the name more generalized. Also that this is not comment-specific.

One example is at profile details, soon enough we might allow markdown in their bio or somewhere else.

**Review by sshanzel — COMMENTED** [review]
Just had some very minor comments, but what a great start on this awesome feature! 🚀 

Super excited about this to be in production! 🚢

File: packages/shared/src/components/comments/CommentMarkdown.tsx (hunk 2 above)
Comment (rebelchris) [reply packages/shared/src/components/comments/CommentMarkdown.tsx:1]: Ah very good point Lee, thanks for this :D

**Comment by rebelchris** [comment]
@sshanzel Would you mind taking another look?

File: packages/shared/src/components/Markdown.tsx [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+export default function Markdown({
+  children,
+}: {
+  children: string;
+}): ReactElement {
+  return (
+    <div
+      className={styles.markdown}
+      dangerouslySetInnerHTML={{ __html: sanitize(children) }}
```
Comment (idoshamun) [inline packages/shared/src/components/Markdown.tsx:13]: Love the fact that you remembered to sanitized the children

File: packages/shared/src/components/Markdown.tsx [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import React, { ReactElement } from 'react';
+import { sanitize } from 'dompurify';
+import styles from './markdown.module.css';
+
+export default function Markdown({
+  children,
+}: {
+  children: string;
```
Comment (idoshamun) [inline packages/shared/src/components/Markdown.tsx:8]: I think that replacing the children's type or using it for the markdown is not a convention. It might be confusing.
Please use a different name.

File: packages/shared/src/components/markdown.module.css [hunk 5]
```diff
@@ -0,0 +1,125 @@
+.markdown {
+  & :where(h1) {
```
Comment (idoshamun) [inline packages/shared/src/components/markdown.module.css:2]: Why do you use `& :where(x)` instead of `& x`?

File: packages/shared/src/components/markdown.module.css [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    @apply rounded-16;
+  }
+  & > :where(:first-child) {
+    @apply mt-0;
+  }
+  & > :where(:last-child) {
+    @apply mb-0;
+  }
+  & :where(pre) :global(:where(.hljs-comment, .hljs-title)) {
+    color: #989498;
```
Comment (idoshamun) [inline packages/shared/src/components/markdown.module.css:63]: Why do you use hard-coded colors here?

File: packages/shared/src/components/modals/DeleteCommentModal.spec.tsx [hunk 7]
```diff
@@ -37,13 +37,6 @@ const renderComponent = (
   );
 };
 
-it('should close modal on cancel', async () => {
```
Comment (idoshamun) [inline packages/shared/src/components/modals/DeleteCommentModal.spec.tsx:40]: Why did you delete this test?

File: packages/shared/src/components/modals/DiscardCommentModal.spec.tsx [hunk 8]
```diff
@@ -20,13 +20,6 @@ const renderComponent = (props: Partial<Props> = {}): RenderResult => {
   return render(<DiscardCommentModal {...defaultProps} {...props} />);
 };
 
-it('should close modal on cancel', async () => {
```
Comment (idoshamun) [inline packages/shared/src/components/modals/DiscardCommentModal.spec.tsx:23]: Why did you delete this test?

**Review by idoshamun — CHANGES_REQUESTED** [review]
I don't see the "Markdown supported" link in the new comment modal as shown in the design.
Otherwise, this seems like an incredible work

File: packages/shared/src/components/markdown.module.css (hunk 5 above)
Comment (rebelchris) [reply packages/shared/src/components/markdown.module.css:2]: More stricter selector in terms of resetting the `:first-child` and `:last-child` and child components.
Also enables us to target multiple at once for instance the `ul, ol` one.

I've taking this approach from Tailwind, after finding issues with resetting  child types.

File: packages/shared/src/components/markdown.module.css (hunk 6 above)
Comment (rebelchris) [reply packages/shared/src/components/markdown.module.css:63]: Good question, just used a default color scheme that was available.
How would you suggest we go about it?

File: packages/shared/src/components/modals/DeleteCommentModal.spec.tsx (hunk 7 above)
Comment (rebelchris) [reply packages/shared/src/components/modals/DeleteCommentModal.spec.tsx:40]: This has been deleted since the cancel button is now replaced with the "Markdown supported" link

File: packages/shared/src/components/markdown.module.css (hunk 6 above)
Comment (idoshamun) [reply packages/shared/src/components/markdown.module.css:63]: You should advise @tsahimatsliah about the right colors to use.
You can send him the default palette and he'll send you back our palette.
It has to support light mode as well

File: packages/shared/src/components/markdown.module.css (hunk 5 above)
Comment (idoshamun) [reply packages/shared/src/components/markdown.module.css:2]: Thanks for explaining :) Just make sure we have a proper browser support or that PostCSS compiles it to something every browser supports

**Review by idoshamun — CHANGES_REQUESTED** [review]
Looks way better! Thanks :)

**Comment by rebelchris** [comment]
Writing a subnote here:
Do not forget to do the API first and run the convert command 👏

**Review by idoshamun — APPROVED** [review]
Sweet!! Once we deploy the API I'll run the script and we can merge this
